docs/design/jet/023-kafka-connect-source-connectors.md

```diff
@@ -189,6 +189,21 @@ There are also nice to have features:
 - Add benchmarks for Kafka connect source
 - Add Soak tests for Kafka connect source
 
+#### Parallelization and scaling
+
+Each node in the cluster creates an instance of `ConnectorWrapper`. 
+Each Jet processor runs one `TaskRunner`, which is a wrapper for Kafka Connect's `SourceTask`s. 
+
+Before the job starts executing we assign `tasks.max` property to total parallelism of the Jet job.
+During the initialization and any reconfiguration requests, a `ConnectorWrapper` asks all associated
+`TaskRunner`s to update their configuration and restart. Each `TaskRunner` gets a configuration
+`conf[processorGlobalIndex]`, where `conf` is a list of configurations returned by Kafka Connect 
+connector. If `conf.size() < processorGlobalIndex`, then such `TaskRunner` becomes a no-op processor.
+
+While the solution is pretty simple, it depends on one assumption - that Kafka Connect connector
+returns stable list, meaning that in given point in time every connector with the same base configuration
+will return exactly the same configuration list for tasks on each member.
```

Comment:
Did you verify this assumption in some implementations?

What about race conditions in the task reconfiguration? Say the following happens:

We run with total parallelism 4, on 2 nodes, connector returns 3 tasks - e.g. x, y, z

p1 -> x
p2 -> y
p3 -> z
p4 -> no-op

The connector asks for reconfiguration, new tasks are: x, a, y, z

p1 -> x
p2 -> a
p3 -> y
p4 -> z

but if e.g. node 2 is reconfigured slightly earlier than node 1, the task y runs on 2 nodes for a short period of time.


Or do you also expect that new tasks are added to the end of the list?

Reply:
I did no assumptions here. In case of reconfiguration every task is restarted (not only those which has new config) and the reading restarts from last seen offset on given partition. 

That being said, user must be aware that duplicates may happen and if user wants e.g. exactly-once, then sink must do the deduplication.

Reply:
For example JdbcSourceConnector uses `ConnectorUtils#groupPartitions` to group tables into partitions (which will end up as separate `SourceTask`s), algorithm is stable for that can I say.

Reply:
> he reading restarts from last seen offset on given partition.

How is the state transferred from node 1 to node 2?

Reply:
In case of:
1) Node down - job is restarted from the snapshot and the snapshot includes all offsets from all nodes (those are merged)
2) Node up:
  a) auto scaling on: job will restart from snapshot as in case of node down
  b) auto scaling off: no changes in the job will be made, kafka tasks also won't be affected by new node